lib/cdo/log_collector.rb

```diff
@@ -0,0 +1,91 @@
+# LogCollector is a simple container that collects errors and important info
+# when executing a process. It can also time block execution and log the result.
+#
+# LogCollector is helpful when we want to
+# - Prevent non-fatal errors from stopping process execution but still want to know about them.
+# - Bubble up combined errors and info from lower levels to higher levels, or to external components
+#   such as HoneyBadger and Slack.
+#
+class LogCollector
+  attr_reader :errors, :logs
+
+  def initialize
+    # List of rescued error objects
+    @errors = []
+    # List of message logs
+    @logs = []
+  end
+
+  # Execute a block and time it.
+  # Save exception if caught, do not re-raise. Caller's flow will continue as normal.
+  #
+  # @param action_name [string] a friendly name of the block being executed
+  def time(action_name = nil)
+    return unless block_given?
+    start_time = Time.now
+
+    yield
+
+    info("#{action_name || 'Unnamed'} action completed without error in"\
+      " #{self.class.get_friendly_time(Time.now - start_time)}."
+    )
+  rescue StandardError => e
+    error("#{action_name || 'Unnamed'} action exited with error in"\
+      " #{self.class.get_friendly_time(Time.now - start_time)}."
+    )
+    record_exception(e)
+  end
+
+  # Execute a block and time it.
+  # Re-raise execption if caught, do not save. This will disrupt the caller's flow.
+  #
+  # @param action_name [string] friendly name for the given block
+  #
+  # @raise [StandardError] error encoutered when executing the given block
+  def time!(action_name = nil)
+    return unless block_given?
+    start_time = Time.now
+
+    yield
+
+    info("#{action_name || 'Unnamed'} action completed without error in"\
+      " #{self.class.get_friendly_time(Time.now - start_time)}."
+    )
+  rescue StandardError => e
+    error("#{action_name || 'Unnamed'} action exited with error in"\
+      " #{self.class.get_friendly_time(Time.now - start_time)}. Exception re-raised!"
+    )
+
+    # To be handled by caller
+    raise e
```

Comment:
Idiomatic Ruby note: [Calling `raise` without arguments will re-raise the last error](https://stackoverflow.com/questions/23770924/reraise-same-exception-after-catching-an-exception-in-ruby), so you could drop the `e` parameter entirely and write:

```ruby
rescue StanardError
  error(...)
  raise
end
```